proposals/src/ledger-replication-to-implement.md

```diff
@@ -37,3 +37,84 @@ transacation proves the validator incorrectly validated a fake storage proof.
 The replicator is rewarded and the validator's staking balance is slashed or
 frozen.
 
+### Storage proof contract logic
+
+Each replicator and validator will have their own storage account. The validator's
+account would be separate from their gossip id similiar to their vote account.
+These should be implemented as two programs one which handles the validator as the keysigner
+and one for the replicator. In that way when the programs reference other accounts, they
+can check the program id to ensure it is a validator or replicator account they are
+referencing.
+
+#### SubmitMiningProof
+```compile_fail
+SubmitMiningProof {
+    slot: u64,
+    sha_state: Hash,
+    signature: Signature,
+};
+keys = [replicator_keypair]
+```
+Replicators create these after mining their stored ledger data for a certain hash value.
+The slot is the end slot of the segment of ledger they are storing, the sha\_state
+the result of the replicator using the hash function to sample their encrypted ledger segment.
+The signature is the signature that was created when they signed a PoH value for the
+current storage epoch. The list of proofs from the current storage epoch should be saved
+in the account state, and then transfered to a list of proofs for the previous epcoh when
+the epoch passes. In a given storage epoch a given replicator should only submit proofs
+for one segment. The program should check that the slot and signature is valid and then store
+a record of this proof being submitted in the current storage epoch.
+
+#### ProofValidation
+```compile_fail
+ProofValidation {
+   proof_mask: Vec<ProofStatus>,
+}
+keys = [validator_keypair, replicator_keypair(s) (unsigned)]
+```
+A validator will submit this transaction to indicate that a set of proofs for a given
+segment are valid/not-valid or skipped where the validator did not look at it. The
+keypairs for the replicators that it looked at should be referenced in the keys so the program
+logic can go to those accounts and see that the proofs are generated in the previous epoch. The
+sampling of the storage proofs should be verified ensuring that the correct proofs are skipped by
+the validator according to the logic outlined in the validator behavior of sampling.
+
+#### ClaimStorageReward
+```compile_fail
+ClaimStorageReward {
+}
+keys = [validator_keypair or replicator_keypair, validator/replicator_keypairs (unsigned)]
+```
+Replicators and validators will use this transaction to get paid tokens from a program state
+where SubmitStorageProof, ProofValidation and ChallengeProofValidations are in a state where
+proofs have been submitted and validated and there are no ChallengeProofValidations referencing
+those proofs. For a validator, it should reference the replicator keypairs to which it has validated
+proofs in the relevant epoch. And for a replicator it should reference validator keypairs for which it
+has validated and wants to be rewarded.
+
+#### ChallengeProofValidation
```

Comment:
oh I dont think I understand this at all. Why do replicators submit fake proofs? shouldn't that be slash-able?

Reply:
They do it to prevent validators from just approving everything they see. The value they use for the hash though is a short value made up by the replicator, then the replicator can show the network the seed value it used to create the hash to catch the validator. In the case it catches the validator, the validator should not be paid for any ProofValidations that it produces in that epoch.

Reply:
Ah I see. So with every set of proofs in an epoch, some of them must be fake so that replicators keep validators in check. Validators need to be sure to keep sampling replicators every now and then too.

Reply:
Yes, that's right.